### PR TITLE
Fix potential exponential behavior in obligation tracking in Must Cal…

### LINKED ISSUE #6566: Potential Exponential Behavior in Obligation Tracking in Must Call Inference
## Description
During the investigation of what was initially suspected to be a non-termination bug in the `url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8` project from the NJR dataset, it was discovered that the issue may actually be exponential growth in complexity rather than non-termination. This behavior occurs due to an unchecked accumulation of obligations to track across multiple control-flow paths when the bailout check `MustCallConsistencyAnalyzer::shouldTrackInvocationResult` is absent.

## Steps to Reproduce
- Run the application on the `url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8` project from the NJR dataset, wit h resourceleak inference enabled. 
    Source: [url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8.zip](https://github.com/typetools/checker-framework/files/15180883/url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8.zip)
- Execute the reproducing unit test `tests/ainfer-resourceleak/non-annotated/FuzzyEngine.java` within the `ainfer-resourceleak` test suite, noting the out-of-memory error that occurs without the proposed modifications, highlighting the unchecked growth in resource consumption.

## Fix
The `MustCallConsistencyAnalyzer` has a bail-out check before tracking obligations using the `shouldTrackInvocationResult` method. The original method was designed primarily for MustCall consistency checks and did not accommodate the unique requirements of resource leak inference.

The original method does not track the invocation result in these cases:

- Constructor Calls: Skips tracking for `this()` or `super()` constructor calls.
- MustCallAlias Annotations: Omits tracking when the method’s return type is `MustCallAlias` and corresponds to an owning field.
- Non-owning Return Types: Avoids tracking for methods annotated with `NotOwning` or methods that have no return type.

For example, consider the following code snippet:

```java
@InheritableMustCall("a")
private class MCASuperClass {
  final @Owning Foo f;

  @MustCallAlias MCASuperClass(@MustCallAlias Foo foo) {
    this.f = foo;
  }

  public void a() {
    f.a();
  }
}

private class MCASuperCall extends MCASuperClass {
  MCASuperCall(Foo foo) {
    super(foo);
  }
}
```
In this scenario, if obligations for the `super(foo)` call are not tracked, then the must-call inference cannot infer that the parameter `Foo foo` and the constructor are resource aliases. This failure results in the inability to infer the `@MustCallAlias` annotation correctly for the `MCASuperCall`, leading to potential resource management issues.
```java
@MustCallAlias MCASuperCall(@MustCallAlias Foo foo) {
  super(foo);
}
```

To rectify this, a new parameter `isMustCallInference` was added to the `shouldTrackInvocationResult` method. This parameter allows the method to adapt its behavior based on the context—skipping tracking where safe in general cases but enforcing it in resource leak scenarios to ensure accurate management of obligations.

## CF Invocation Environment
- Ubuntu 20.04
- JDK 11
- Flags
  ```bash
  -processor org.checkerframework.checker.resourceleak.ResourceLeakChecker \
  -Adetailedmsgtext \
  -Aajava=$PROJECT_PATH/wpi-out \
  -Ainfer=ajava \
  -Awarns \
  -AshowPrefixInWarningMessages \
  -AdisableReturnsReceiver \
  ```

### PR DESCRIPTION
**Pull request by iamsanjaymalakar**
…l Inference; fixes #6566

### CODE REVIEW

File: checker/tests/ainfer-resourceleak/non-annotated/FuzzyEngine.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+ * The test case reveals issues in the url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8
+ * project from the NJR dataset, where the lack of the
+ * MustCallConsistencyAnalyzer::shouldTrackInvocationResult bailout check leads to exponential
+ * growth in complexity. This results from an unchecked accumulation of obligations to track across
+ * multiple control-flow paths, originally suspected to cause non-termination.
+ */
+import java.util.ArrayList;
+import java.util.List;
+
+public class FuzzyEngine {
```
Comment (msridhar) [inline checker/tests/ainfer-resourceleak/non-annotated/FuzzyEngine.java:11]: Just checking, did you confirm that without your fix, you see the exponential behavior on this input when running the inference regression tests?

**Review by msridhar — DISMISSED** [review]
Looks great, thanks for the fix!

File: checker/tests/ainfer-resourceleak/non-annotated/FuzzyEngine.java (hunk 1 above)
Comment (iamsanjaymalakar) [reply checker/tests/ainfer-resourceleak/non-annotated/FuzzyEngine.java:11]: Yes, and without the fix I was encountering OOM in my machine.

image

**Comment by msridhar** [comment]
@mernst does the CI failure on this PR look like flakiness to you?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 2]
```diff
@@ -921,9 +921,11 @@ private boolean areSame(JavaExpression target, @Nullable JavaExpression enclosin
    * @param obligations the current set of Obligations, which may be side-effected
    * @param node the invocation node to check; must be {@link MethodInvocationNode} or {@link
    *     ObjectCreationNode}
+   * @param isMustCallInference true if this call is invoked as part of a MustCall inference
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:924]: a "call" cannot be "invoked".  drop "is invoked", or change "method" to "call"

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 3]
```diff
@@ -939,8 +941,12 @@ private boolean shouldTrackInvocationResult(Set<Obligation> obligations, Node no
     // Now callTree.getKind() == Tree.Kind.METHOD_INVOCATION.
     MethodInvocationTree methodInvokeTree = (MethodInvocationTree) callTree;
 
-    if (TreeUtils.isSuperConstructorCall(methodInvokeTree)
-        || TreeUtils.isThisConstructorCall(methodInvokeTree)) {
+    // For must call inference, we do not want to bail-out on tracking the obligations for 'this()'
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:944]: "bail out" is two words.

**Review by mernst — DISMISSED** [review]

**Comment by mernst** [comment]
> @mernst does the CI failure on this PR look like flakiness to you?

Yes.  Errors in "junit_jdk11" and "junit_jdk17" without an error in "junit_jdk21" are almost always CI flakiness.  Suzanne and I have not yet figured it out, but I can click the "rerun" button.  (For now, there are two minor typos to fix.)

**Review by msridhar — APPROVED** [review]

**Pull request merged**